soccer/src/referee/ExternalReferee.cpp

```diff
@@ -0,0 +1,223 @@
+    try {
+        _asio_socket.bind(listen_endpoint);
+    } catch (const boost::system::system_error& e) {
+        throw std::runtime_error("Failed to bind to shared referee port");
+    }
+    // Join multicast group
+    const boost::asio::ip::address multicast_address =
+        boost::asio::ip::address::from_string(RefereeAddress);
+    _asio_socket.set_option(
+        boost::asio::ip::multicast::join_group(multicast_address));
+}
+
+void ExternalReferee::run() {
+    _io_service.poll();
+    BallState state;
+    spin_kick_detector(state);
+}
+
+void ExternalReferee::handle_command(SSL_Referee::Command command) {
+    switch (command) {
+        case SSL_Referee::HALT:
+            halt();
+            break;
+        case SSL_Referee::STOP:
+            stop();
+            break;
+        case SSL_Referee::NORMAL_START:
+            break;
+        case SSL_Referee::FORCE_START:
+            play();
+            break;
+        case SSL_Referee::PREPARE_KICKOFF_YELLOW:
+            restart(GameState::Restart::Kickoff, false);
+            setup();
+            break;
+        case SSL_Referee::PREPARE_KICKOFF_BLUE:
+            restart(GameState::Restart::Kickoff, true);
+            setup();
+            break;
+        case SSL_Referee::PREPARE_PENALTY_YELLOW:
+            restart(GameState::Restart::Penalty, false);
+            setup();
+            break;
+        case SSL_Referee::PREPARE_PENALTY_BLUE:
+            restart(GameState::Restart::Penalty, true);
+            setup();
+            break;
+        case SSL_Referee::DIRECT_FREE_YELLOW:
+            restart(GameState::Restart::Direct, false);
+            break;
+        case SSL_Referee::DIRECT_FREE_BLUE:
+            restart(GameState::Restart::Direct, true);
+            break;
+        case SSL_Referee::INDIRECT_FREE_YELLOW:
+            restart(GameState::Restart::Indirect, false);
+            break;
+        case SSL_Referee::INDIRECT_FREE_BLUE:
+            restart(GameState::Restart::Indirect, true);
+            break;
+        case SSL_Referee::TIMEOUT_YELLOW:
+        case SSL_Referee::TIMEOUT_BLUE:
+            halt();
+            break;
+        case SSL_Referee::GOAL_YELLOW:
+        case SSL_Referee::GOAL_BLUE:
+            break;
+        case SSL_Referee::BALL_PLACEMENT_YELLOW:
+            // TODO(Kyle): ball placement point
+            ball_placement(Geometry2d::Point{}, false);
+            stop();
+            break;
+        case SSL_Referee::BALL_PLACEMENT_BLUE:
+            // TODO(Kyle): ball placement point
+            ball_placement(Geometry2d::Point{}, true);
+            stop();
+            break;
+    }
+}
+
+void spin_external_referee() {
+    auto external_ref = std::make_shared<ExternalReferee>();
+
+    rclcpp::executors::SingleThreadedExecutor executor;
+    executor.add_node(external_ref);
+
+    while (rclcpp::ok()) {
+        executor.spin_some(std::chrono::milliseconds(5));
```

Comment:
fyi spin_some doesn't work in the way you expect it to work. I've tested this, and it basically only calls 1 callback no matter how much time you give it. For referee though I think it should be fine.

Reply:
IIRC it only processes the currently-ready items (ready = queued at the start of spin_some) but will exit before this time if it's already processed all of those items:

> Complete all available queued work without blocking.
> This function can be overridden. The default implementation is suitable for a single-threaded model of execution. Adding subscriptions, timers, services, etc. with blocking callbacks will cause this function to block (which may have unintended consequences).

Reply:
I mean like I've tested out creating a publisher publishing at 100hz and a subscriber using `spin_some` at 10hz, and the subscriber only gets 1 message every time.